[WIP] add frequency API and compute coffeine API

dengemann opened the pull request:
Closes #41 #42 #11 #28  – API included in both notebook and in library.

This PR implements 3 main ideas:

1. Have a function to get frequencies / bands from pre-defined sets in the literature
2. Have a documented & simple function for moving from 4D covariance arrays to the coffeine data frame (low-level).
3. Have a main function that beams us from M/EEG time-series to our coffeine data frame – with some sugar API for doing a few things in one function call (including 1 & 2).

All this should make it far easier for us to express & perform data analyses, reducing biolerplate & cognitive burden. And newcomers should be happier as well.

dengemann commented on coffeine/power_features.py:
```diff
@@ -53,6 +55,214 @@ def _compute_cospectral_covs(epochs, n_fft, n_overlap, fmin, fmax, fs):
+    return frequency_bands
+
+
+def make_coffeine_df(C: np.ndarray,
```
@apmellot @antoinecollas @apmellot @DavidSabbagh @hubertjb @bmalezieux 

A small but important & overdue function :) What do you think about its naming? I think I am in principle happy with it. Other intuitive options would be to call it `make_coffeine_data_frame` or `make_covariance_data_frame`.  I would be OK with making it explicit (`data_frame` rather than `df`). Slight preference for coffeine `make_coffeine_data_frame` for branding and energizing weirdness. I understand `make_coffeine_data_frame`  could be more boring and predictable (in the dispassionate scientific sense), which would be not bad either.

Any thoughts? (for the bigger picture, please take minute to zoom out and look at the entire PR / the linked issues).

apmellot replied:
I prefer `make_coffeine_data_frame` to be consistent with the function `make_coffeine`

agramfort commented:
I would go with make_covariance_data_frame or make_filter_bank_covariance

dengemann commented:
Now would be a good moment for a first round of reviews at the conceptual/strategic/API level. Once the roadmap / idea is clear I‘ll work on new unit tests.

dengemann commented:
> Closes #41 #42 – API included in both notebook and in library.
> 
> This PR implements 3 main ideas:
> 
> 1. Have a function to get frequencies / bands from pre-defined sets in the literature
> 2. Have a documented & simple function for moving from 4D covariance arrays to the coffeine data frame (low-level).
> 3. Have a main function that beams us from M/EEG time-series to our coffeine data frame – with some sugar API for doing a few things in one function call (including 1 & 2).
> 
> All this should make it far easier for us to express & perform data analyses, reducing biolerplate & cognitive burden. And newcomers should be happier as well.

Also note - I put some thoughts into this with possible future developments in mind (alternative frequency schemes) and alternative feature computation options, including additional covariance estimators (for which we should best update the MNE-Python API, estimator by estimator, as this is easy, specific and gets us all benefits of MNE‘s capability of handling multiple sensor/electrode types + data rank).

dengemann commented:
ok seems like we‘ve got docs 😸

agramfort commented:
indeed
https://github.com/coffeine-labs/coffeine/suites/14376119098/artifacts/810044724
!

> Message ID: ***@***.***>
>

agramfort commented:
works for me

> Message ID: ***@***.***>
>

dengemann commented:
ok - we've never been that far!

- implemented comments
- added missing doc strings
- added typehints where appropriate
- upgraded testing worlflow to python=3.9 to accomodate typhints

it feels quite big now. unless anyone protests I'd merge this to see how docs / website look.
we have to make another PRs to add more tests and implement suggestions by Alex on Kernel code.

cc @apmellot @agramfort

agramfort reviewed (changes requested):
you should not commit anything from the generated folder. Please remove it and add it to .gitignore

dengemann commented:
... one more test missing :)

dengemann commented:
Ok, I'm merging 🚀 🎸 – if you have complaints please open an issue / PR 😸 🐙
